src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs

```diff
@@ -92,19 +92,9 @@ private static bool DirectoryExists(string fullPath)
             return ((fileinfo.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFDIR);
         }
 
-        /// <summary>Opens a SafeFileHandle for a file descriptor created by a provided delegate.</summary>
-        /// <param name="fdFunc">
-        /// The function that creates the file descriptor. Returns the file descriptor on success, or an invalid
-        /// file descriptor on error with Marshal.GetLastWin32Error() set to the error code.
-        /// </param>
-        /// <returns>The created SafeFileHandle.</returns>
-        internal static SafeFileHandle Open(Func<SafeFileHandle> fdFunc)
-        {
-            SafeFileHandle handle = Interop.CheckIo(fdFunc());
-
-            Debug.Assert(!handle.IsInvalid, "File descriptor is invalid");
-            return handle;
-        }
+        // Each thread will have its own copy. This locks the file and avoids race conditions if the handle had the last error.
```

Comment:
> This locks the file

I am not sure what you mean. This does not lock the file.

Reply:
Are you trying to handle a scenario where ReleaseHandle() on a particular SafeFileHandle is called by two different threads, and you want the one that errors to be the one that gets the error when it calls Dispose()? If so is it documented that SafeFileHandle can be used in this way? The docs say "The ReleaseHandle method is guaranteed to be called only once and only if the handle is valid as defined by the IsInvalid property."

Why can't it be a regular field, as `handle` is?

Reply:
`ReleaseHandle` is guaranteed to be called just once. `Dispose` can be called multiple times, from different threads.

If this was a regular field, we would not be able to tell whether the error happened on this thread or whether the error happened on a different thread. Moreover, the last error is `Nullable<int>` currently that cannot be updated atomically, so we could end up in situations where the nullable is non-null, but the error is still not set.

It is certainly possible to implement this using a regular field. It will have different behavior in corner cases, and it needs a special care to handle all possible race conditions.

Reply:
Fixed the comment.

Reply:
I see. I assume Dispose() always calls ReleaseHandle() if it has not already been called. So it is possible for another thread to Dispose this SafeFileHandle, and ignore the exception, and then FileStream's Dispose would not be aware. Are we OK with that -- perhaps because we don't expect it to ever happen? As opposed to exposing the ErrorInfo to every caller of Dispose(), even if they are not the first.

I'm fine with that, just wanted to undestand what we wanted.